virttest/test_setup.py

```diff
@@ -97,6 +100,90 @@ class PolkitConfigCleanupError(PolkitConfigError):
     pass
 
 
+class Setuper(object):
+
+    """
+    Virtual base abstraction of setuper.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, test, params, env):
+        """
+        Initializes the setuper.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self.test = test
+        self.params = params
+        self.env = env
+
+    @abstractmethod
+    def setup(self):
+        """Setup procedure."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def cleanup(self):
+        """Cleanup procedure."""
+        raise NotImplementedError
+
+
+class SetuperMgr(object):
+
+    """
+    Setuper Manager implementation.
+
+    The instance can help do the setup stuff before test started and
+    do the cleanup stuff after test finished. This setup-cleanup
+    combined stuff will be performed in LIFO order.
+    """
+
+    def __init__(self, test, params, env):
+        """
+        Initializes the setuper manager.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self._stp_args = (test, params, env)
+        self._reg_stps = []
+        self._setupers = []
+
+    def register(self, stp_cls):
+        """
+        Register the given setuper class to the manager.
+
+        :param stp_cls: Setuper class.
+        """
+        self._reg_stps.append(stp_cls)
+
+    def do_setup(self):
+        """Do setup stuff."""
+        for stp_cls in self._reg_stps:
+            stp = stp_cls(*self._stp_args)
+            stp.setup()
+            self._setupers.append(stp)
+
+    def do_cleanup(self):
+        """
+        Do cleanup stuff.
+
+        :return: Errors occurred in cleanup procedures.
+        """
+        err = []
+        while self._setupers:
+            try:
+                self._setupers.pop().cleanup()
```

Comment:
This means in case of some `setup` failure other cleanups might not be executed. I think it'd be better to always execute `cleanup`, because some cleanups are not related to setup like terminating all vms and killing background processes (can be solved by registering them first) and some contain setup (which can fail) but still should be always executed (eg. postprocess images; here we can't really implement them using this as earlier failed setup causes this cleanup not to be executed).

Reply:
It sounds like a setuper only has one of the `setup` and `cleanup` method, so my question is do we really have such the case, I think almost every single `setup` should have a `cleanup`, vice versa. For example, we can pair `preprocess_image` and `postprocess_image` up, `preprocess_image` is `setup` and `postprocess_image` is `cleanup`.